src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2160,6 +2160,14 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
                     {
                         SymbolDistinguisher distinguisher = new SymbolDistinguisher(this.Compilation, lambdaParameterType, delegateParameterType);
 
+                        // Report the error beforehand to preserve previous behavior
+                        if (!hasTypeMismatch)
+                        {
+                            hasTypeMismatch = true;
+                            // Cannot convert {0} to type '{1}' because the parameter types do not match the delegate parameter types
+                            Error(diagnostics, ErrorCode.ERR_CantConvAnonMethParams, syntax, id, targetType);
+                        }
```

Comment:
would it make more sense to just move the ref-kind checks first, and not report the type issue if there's a refkind issue?  

That seems sensible to me.  Tell the user first "your refkind is way off, you need to fix that no matter what".  Then, once refkinds are good, do the equality check.

Reply:
So you're saying you'd rather postpone reporting all type inequality errors until ref-kind equality errors are resolved first, which I like as an idea

Reply:
yes.  it seems to be similar to what you're trying here, but cleaner and more intuitive (To me).